sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1085 @@
+
+	// encode CertID, used as a key in the cache
+	encodedCertID := &certIDKey{
+		r.HashAlgorithm,
+		base64.StdEncoding.EncodeToString(r.IssuerNameHash),
+		base64.StdEncoding.EncodeToString(r.IssuerKeyHash),
+		r.SerialNumber.String(),
+	}
+	return encodedCertID, &ocspStatus{
+		code: ocspSuccess,
+	}
+}
+
+func (c *Client) encodeCertIDKey(certIDKeyBase64 string) (*certIDKey, error) {
+	r, err := base64.StdEncoding.DecodeString(certIDKeyBase64)
+	if err != nil {
+		return nil, err
+	}
+	var cid certID
+	rest, err := asn1.Unmarshal(r, &cid)
+	if err != nil {
+		// error in parsing
+		return nil, err
+	}
+	if len(rest) > 0 {
+		// extra bytes to the end
+		return nil, err
+	}
+	return &certIDKey{
+		c.getHashAlgorithmFromOID(cid.HashAlgorithm),
+		base64.StdEncoding.EncodeToString(cid.NameHash),
+		base64.StdEncoding.EncodeToString(cid.IssuerKeyHash),
+		cid.SerialNumber.String(),
+	}, nil
+}
+
+func decodeCertIDKey(k *certIDKey) (string, error) {
+	serialNumber := new(big.Int)
+	serialNumber.SetString(k.SerialNumber, 10)
+	nameHash, err := base64.StdEncoding.DecodeString(k.NameHash)
+	if err != nil {
+		return "", err
+	}
+	issuerKeyHash, err := base64.StdEncoding.DecodeString(k.IssuerKeyHash)
+	if err != nil {
+		return "", err
+	}
+	hashAlgoOid, err := getOIDFromHashAlgorithm(k.HashAlgorithm)
+	if err != nil {
+		return "", err
+	}
+	encodedCertID, err := asn1.Marshal(certID{
+		pkix.AlgorithmIdentifier{
+			Algorithm:  hashAlgoOid,
+			Parameters: asn1.RawValue{Tag: 5 /* ASN.1 NULL */},
+		},
+		nameHash,
+		issuerKeyHash,
+		serialNumber,
+	})
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(encodedCertID), nil
+}
+
+func (c *Client) checkOCSPResponseCache(encodedCertID *certIDKey, subject, issuer *x509.Certificate) (*ocspStatus, error) {
+	c.ocspResponseCacheLock.RLock()
+	var cacheValue *ocspCachedResponse
+	v, ok := c.ocspResponseCache.Get(*encodedCertID)
+	if ok {
+		cacheValue = v.(*ocspCachedResponse)
+	}
+	c.ocspResponseCacheLock.RUnlock()
+
+	status, err := c.extractOCSPCacheResponseValue(cacheValue, subject, issuer)
+	if err != nil {
+		return nil, err
+	}
+	if !isValidOCSPStatus(status.code) {
+		c.deleteOCSPCache(encodedCertID)
+	}
+	return status, err
+}
+
+func (c *Client) deleteOCSPCache(encodedCertID *certIDKey) {
+	c.ocspResponseCacheLock.Lock()
+	c.ocspResponseCache.Remove(*encodedCertID)
+	c.cacheUpdated = true
+	c.ocspResponseCacheLock.Unlock()
+}
+
+func validateOCSP(ocspRes *ocsp.Response) (*ocspStatus, error) {
+	curTime := time.Now()
+
+	if ocspRes == nil {
+		return nil, errors.New("OCSP Response is nil")
+	}
+	if !isInValidityRange(curTime, ocspRes.NextUpdate) {
+		return &ocspStatus{
+			code: ocspInvalidValidity,
+			err:  fmt.Errorf("invalid validity: producedAt: %v, thisUpdate: %v, nextUpdate: %v", ocspRes.ProducedAt, ocspRes.ThisUpdate, ocspRes.NextUpdate),
+		}, nil
+	}
+	return returnOCSPStatus(ocspRes), nil
+}
+
+func returnOCSPStatus(ocspRes *ocsp.Response) *ocspStatus {
```

Comment:
Seems weird that we're translating from one internal form to another internal form?

Reply:
Is this possibly for their file caching?

Reply:
Yeah, wondering if perhaps it is that. Then they're stable internal identifiers rather than external identifiers that might potentially change and break their cached entries. 

But still, I don't think Go upstream would make that type of constants change, except maybe if they import x/ocsp into the main series.